DATA PREPROCESSING AND FEATURE ENGINEERING IN MACHINE LEARNING

1.Data Exploration and Preprocessing:

Loading the dataset:

In [4]:
import pandas as pd

In [5]:
df=pd.read_csv("C:\\DataScience25\\official_assignments\\EDA2\\EDA2\\adult_with_headers.csv")

In [6]:
df

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

In [7]:
df.shape

(32561, 15)

In [8]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'income'],
      dtype='object')

Looking for any missing values:

In [10]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

There are no missing values in the dataset.

Finding data type of the features:

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


Summary Statistics:

In [15]:
df.describe()

age        fnlwgt  education_num  capital_gain  capital_loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours_per_week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

Mean age is approximately 38.5 years. The wide range from 17 to 90 suggests the presence of potential outliers in the age column.

Education levels (in numerical form) show a possible presence of outliers, with values ranging from 1 to 16. The mean is around 10.

Capital gain and loss columns are highly skewed. Most values are 0, but a few extremely high values (gain up to 99999, loss up to 4356) cause large spread and indicate clear presence of outliers.

People work on average around 40 hours per week, but the minimum value of 1 and maximum of 99 indicate potential outliers in this column.



Apply scaling techniques to numerical features:

In [18]:
df.select_dtypes('int').columns

Index(['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss',
       'hours_per_week'],
      dtype='object')

In [19]:
from sklearn.preprocessing import StandardScaler

ss=StandardScaler()
scaled_data=ss.fit_transform(df[['age']])
df['age_scaled'] = scaled_data


In [20]:
df=df.drop('age',axis=1)

In [21]:
from sklearn.preprocessing import StandardScaler

ss=StandardScaler()
scaled_data1=ss.fit_transform(df[['fnlwgt']])
df['fnlwgt_scaled'] = scaled_data1


In [22]:
from sklearn.preprocessing import StandardScaler

ss=StandardScaler()
scaled_data2=ss.fit_transform(df[['education_num']])
df['education_num_scaled'] = scaled_data2

In [23]:
df=df.drop('education_num',axis=1)

In [24]:
from sklearn.preprocessing import StandardScaler

ss=StandardScaler()
scaled_data3=ss.fit_transform(df[['capital_gain']])
df['capital_gain_scaled'] = scaled_data3

In [25]:
df=df.drop('capital_gain',axis=1)

In [26]:
from sklearn.preprocessing import StandardScaler

ss=StandardScaler()
scaled_data4=ss.fit_transform(df[['capital_loss']])
df['capital_loss_scaled'] = scaled_data4

In [27]:
df=df.drop('capital_loss',axis=1)

In [28]:
from sklearn.preprocessing import StandardScaler

ss=StandardScaler()
scaled_data5=ss.fit_transform(df[['hours_per_week']])
df['hours_per_week_scaled'] = scaled_data5

In [29]:
df=df.drop('hours_per_week',axis=1)

In [30]:
df

workclass  fnlwgt    education       marital_status  \
0              State-gov   77516    Bachelors        Never-married   
1       Self-emp-not-inc   83311    Bachelors   Married-civ-spouse   
2                Private  215646      HS-grad             Divorced   
3                Private  234721         11th   Married-civ-spouse   
4                Private  338409    Bachelors   Married-civ-spouse   
...                  ...     ...          ...                  ...   
32556            Private  257302   Assoc-acdm   Married-civ-spouse   
32557            Private  154374      HS-grad   Married-civ-spouse   
32558            Private  151910      HS-grad              Widowed   
32559            Private  201490      HS-grad        Never-married   
32560       Self-emp-inc  287927      HS-grad   Married-civ-spouse   

               occupation    relationship    race      sex  native_country  \
0            Adm-clerical   Not-in-family   White     Male   United-States   
1         Exec-managerial         Husband   White     Male   United-States   
2       Handlers-cleaners   Not-in-family   White     Male   United-States   
3       Handlers-cleaners         Husband   Black     Male   United-States   
4          Prof-specialty            Wife   Black   Female            Cuba   
...                   ...             ...     ...      ...             ...   
32556        Tech-support            Wife   White   Female   United-States   
32557   Machine-op-inspct         Husband   White     Male   United-States   
32558        Adm-clerical       Unmarried   White   Female   United-States   
32559        Adm-clerical       Own-child   White     Male   United-States   
32560     Exec-managerial            Wife   White   Female   United-States   

       income  age_scaled  fnlwgt_scaled  education_num_scaled  \
0       <=50K    0.030671      -1.063611              1.134739   
1       <=50K    0.837109      -1.008707              1.134739   
2       <=50K   -0.042642       0.245079             -0.420060   
3       <=50K    1.057047       0.425801             -1.197459   
4       <=50K   -0.775768       1.408176              1.134739   
...       ...         ...            ...                   ...   
32556   <=50K   -0.849080       0.639741              0.746039   
32557    >50K    0.103983      -0.335433             -0.420060   
32558   <=50K    1.423610      -0.358777             -0.420060   
32559   <=50K   -1.215643       0.110960             -0.420060   
32560    >50K    0.983734       0.929893             -0.420060   

       capital_gain_scaled  capital_loss_scaled  hours_per_week_scaled  
0                 0.148453             -0.21666              -0.035429  
1                -0.145920             -0.21666              -2.222153  
2                -0.145920             -0.21666              -0.035429  
3                -0.145920             -0.21666              -0.035429  
4                -0.145920             -0.21666              -0.035429  
...                    ...                  ...                    ...  
32556            -0.145920             -0.21666              -0.197409  
32557            -0.145920             -0.21666              -0.035429  
32558            -0.145920             -0.21666              -0.035429  
32559            -0.145920             -0.21666              -1.655225  
32560             1.888424             -0.21666              -0.035429  

[32561 rows x 16 columns]

Standard Scaling was used because the dataset contains outliers (e.g., in capital_gain, fnlwgt), and Min-Max Scaling would compress the majority of values near zero. Standard Scaling handles such spread better by using mean and standard deviation, making it more suitable for this dataset.

Discuss the scenarios where each scaling technique is preferred and why.

- Standard Scaling:

Used When:
   - Data is normally distributed or close to normal
   - Algorithms assume data is centered around 0

Why?:
   - Brings data to a common scale with mean = 0 and std = 1
   - Helps in maintaining the shape of the distribution



- Min Max Sacling:

Used When:
   - You need all features in [0, 1] range
   - Dataset doesn’t contain extreme outliers
   - Used in algorithms sensitive to feature scale or neural networks

Why?:
   - Ensures all features contribute equally during model training
   - Neural networks converge faster with values between 0 and 1


2.Encoding Techniques:

In [36]:
df.select_dtypes(include='object').nunique()


workclass          9
education         16
marital_status     7
occupation        15
relationship       6
race               5
sex                2
native_country    42
income             2
dtype: int64

Apply One-Hot Encoding to categorical variables with less than 5 categories.

Will be appyling One Hot Encoding on:
- Race
- Sex
- Income

In [39]:
one_hot_cols = ['sex', 'income', 'race']
df_encoded = pd.get_dummies(df, columns=one_hot_cols, drop_first=True)

In [40]:
df_encoded

workclass  fnlwgt    education       marital_status  \
0              State-gov   77516    Bachelors        Never-married   
1       Self-emp-not-inc   83311    Bachelors   Married-civ-spouse   
2                Private  215646      HS-grad             Divorced   
3                Private  234721         11th   Married-civ-spouse   
4                Private  338409    Bachelors   Married-civ-spouse   
...                  ...     ...          ...                  ...   
32556            Private  257302   Assoc-acdm   Married-civ-spouse   
32557            Private  154374      HS-grad   Married-civ-spouse   
32558            Private  151910      HS-grad              Widowed   
32559            Private  201490      HS-grad        Never-married   
32560       Self-emp-inc  287927      HS-grad   Married-civ-spouse   

               occupation    relationship  native_country  age_scaled  \
0            Adm-clerical   Not-in-family   United-States    0.030671   
1         Exec-managerial         Husband   United-States    0.837109   
2       Handlers-cleaners   Not-in-family   United-States   -0.042642   
3       Handlers-cleaners         Husband   United-States    1.057047   
4          Prof-specialty            Wife            Cuba   -0.775768   
...                   ...             ...             ...         ...   
32556        Tech-support            Wife   United-States   -0.849080   
32557   Machine-op-inspct         Husband   United-States    0.103983   
32558        Adm-clerical       Unmarried   United-States    1.423610   
32559        Adm-clerical       Own-child   United-States   -1.215643   
32560     Exec-managerial            Wife   United-States    0.983734   

       fnlwgt_scaled  education_num_scaled  capital_gain_scaled  \
0          -1.063611              1.134739             0.148453   
1          -1.008707              1.134739            -0.145920   
2           0.245079             -0.420060            -0.145920   
3           0.425801             -1.197459            -0.145920   
4           1.408176              1.134739            -0.145920   
...              ...                   ...                  ...   
32556       0.639741              0.746039            -0.145920   
32557      -0.335433             -0.420060            -0.145920   
32558      -0.358777             -0.420060            -0.145920   
32559       0.110960             -0.420060            -0.145920   
32560       0.929893             -0.420060             1.888424   

       capital_loss_scaled  hours_per_week_scaled  sex_ Male  income_ >50K  \
0                 -0.21666              -0.035429       True         False   
1                 -0.21666              -2.222153       True         False   
2                 -0.21666              -0.035429       True         False   
3                 -0.21666              -0.035429       True         False   
4                 -0.21666              -0.035429      False         False   
...                    ...                    ...        ...           ...   
32556             -0.21666              -0.197409      False         False   
32557             -0.21666              -0.035429       True          True   
32558             -0.21666              -0.035429      False         False   
32559             -0.21666              -1.655225       True         False   
32560             -0.21666              -0.035429      False          True   

       race_ Asian-Pac-Islander  race_ Black  race_ Other  race_ White  
0                         False        False        False         True  
1                         False        False        False         True  
2                         False        False        False         True  
3                         False         True        False        False  
4                         False         True        False        False  
...                         ...          ...          ...          ...  
32556                     False        False

Use Label Encoding for categorical variables with more than 5 categories.

Will be appyling One Hot Encoding on:
- Workclass
- education
- marital_status	
- occupation
- relationship
- native_country

In [43]:
from sklearn.preprocessing import LabelEncoder

label_cols = ['relationship', 'marital_status', 'workclass', 'occupation', 'education', 'native_country']

le = LabelEncoder()
for col in label_cols:
    df_encoded[col] = le.fit_transform(df_encoded[col])

In [44]:
df_encoded

workclass  fnlwgt  education  marital_status  occupation  relationship  \
0              7   77516          9               4           1             1   
1              6   83311          9               2           4             0   
2              4  215646         11               0           6             1   
3              4  234721          1               2           6             0   
4              4  338409          9               2          10             5   
...          ...     ...        ...             ...         ...           ...   
32556          4  257302          7               2          13             5   
32557          4  154374         11               2           7             0   
32558          4  151910         11               6           1             4   
32559          4  201490         11               4           1             3   
32560          5  287927         11               2           4             5   

       native_country  age_scaled  fnlwgt_scaled  education_num_scaled  \
0                  39    0.030671      -1.063611              1.134739   
1                  39    0.837109      -1.008707              1.134739   
2                  39   -0.042642       0.245079             -0.420060   
3                  39    1.057047       0.425801             -1.197459   
4                   5   -0.775768       1.408176              1.134739   
...               ...         ...            ...                   ...   
32556              39   -0.849080       0.639741              0.746039   
32557              39    0.103983      -0.335433             -0.420060   
32558              39    1.423610      -0.358777             -0.420060   
32559              39   -1.215643       0.110960             -0.420060   
32560              39    0.983734       0.929893             -0.420060   

       capital_gain_scaled  capital_loss_scaled  hours_per_week_scaled  \
0                 0.148453             -0.21666              -0.035429   
1                -0.145920             -0.21666              -2.222153   
2                -0.145920             -0.21666              -0.035429   
3                -0.145920             -0.21666              -0.035429   
4                -0.145920             -0.21666              -0.035429   
...                    ...                  ...                    ...   
32556            -0.145920             -0.21666              -0.197409   
32557            -0.145920             -0.21666              -0.035429   
32558            -0.145920             -0.21666              -0.035429   
32559            -0.145920             -0.21666              -1.655225   
32560             1.888424             -0.21666              -0.035429   

       sex_ Male  income_ >50K  race_ Asian-Pac-Islander  race_ Black  \
0           True         False                     False        False   
1           True         False                     False        False   
2           True         False                     False        False   
3           True         False                     False         True   
4          False         False                     False         True   
...          ...           ...                       ...          ...   
32556      False         False                     False        False   
32557       True          True                     False        False   
32558      False         False                     False        False   
32559       True         False                     False        False   
32560      False          True                     False        False   

       race_ Other  race_ White  
0            False         True  
1            False         True  
2            False         True  
3            False        False  
4            False        False  
...            ...          ...  
32556        False         True  
32557        False         True  
32558        False         True  
32559        False         True  
32560

Discuss the pros and cons of One-Hot Encoding and Label Encoding.

One-Hot Encoding:

- Pros:
  - Doesn’t assume any order or priority among categories.
  - Works well with algorithms that don’t handle categorical data directly.

- Cons:
  - Creates many columns → leads to high dimensionality.
  - Increases memory usage and training time.


Label Encoding:

- Pros:
  - Very simple and memory efficient.
  - Keeps the data as a single column.
  - Useful when categories have ordinal meaning.

- Cons:
  - Implies ordinal relationship even if not true.
  - Can mislead algorithms like linear regression, KNN, or SVM.

3.Feature Engineering:

Create at least 2 new features that could be beneficial for the model. Explain the rationale behind your choices.

In [52]:
# Step 1: Create new features on original df
df['is_senior'] = df['age_scaled'].apply(lambda x: 1 if x >= 60 else 0)
df['capital_net'] = df['capital_gain_scaled'] - df['capital_loss_scaled']

# Step 2: Add to encoded df
df_encoded['is_senior'] = df['is_senior']
df_encoded['capital_net'] = df['capital_net']



Created 2 new features:

- is_senior:     
Senior citizens (60+) may have different working hours, income status, or capital gains. This feature helps model detect income pattern differences by age group.

- capital_net:
Instead of treating capital_gain and capital_loss separately, capital_net gives a clear net value of capital flow, which simplifies learning for the model.

In [54]:
import numpy as np

df_encoded['capital_gain_log'] = np.log1p(df['capital_gain_scaled'])


Apply a transformation (e.g., log transformation) to at least one skewed numerical feature and justify your choice.

In [56]:
df_encoded

workclass  fnlwgt  education  marital_status  occupation  relationship  \
0              7   77516          9               4           1             1   
1              6   83311          9               2           4             0   
2              4  215646         11               0           6             1   
3              4  234721          1               2           6             0   
4              4  338409          9               2          10             5   
...          ...     ...        ...             ...         ...           ...   
32556          4  257302          7               2          13             5   
32557          4  154374         11               2           7             0   
32558          4  151910         11               6           1             4   
32559          4  201490         11               4           1             3   
32560          5  287927         11               2           4             5   

       native_country  age_scaled  fnlwgt_scaled  education_num_scaled  ...  \
0                  39    0.030671      -1.063611              1.134739  ...   
1                  39    0.837109      -1.008707              1.134739  ...   
2                  39   -0.042642       0.245079             -0.420060  ...   
3                  39    1.057047       0.425801             -1.197459  ...   
4                   5   -0.775768       1.408176              1.134739  ...   
...               ...         ...            ...                   ...  ...   
32556              39   -0.849080       0.639741              0.746039  ...   
32557              39    0.103983      -0.335433             -0.420060  ...   
32558              39    1.423610      -0.358777             -0.420060  ...   
32559              39   -1.215643       0.110960             -0.420060  ...   
32560              39    0.983734       0.929893             -0.420060  ...   

       hours_per_week_scaled  sex_ Male  income_ >50K  \
0                  -0.035429       True         False   
1                  -2.222153       True         False   
2                  -0.035429       True         False   
3                  -0.035429       True         False   
4                  -0.035429      False         False   
...                      ...        ...           ...   
32556              -0.197409      False         False   
32557              -0.035429       True          True   
32558              -0.035429      False         False   
32559              -1.655225       True         False   
32560              -0.035429      False          True   

       race_ Asian-Pac-Islander  race_ Black  race_ Other  race_ White  \
0                         False        False        False         True   
1                         False        False        False         True   
2                         False        False        False         True   
3                         False         True        False        False   
4                         False         True        False        False   
...                         ...          ...          ...          ...   
32556                     False        False        False         True   
32557                     False        False        False         True   
32558                     False        False        False         True   
32559                     False        False        False         True   
32560                     False        False        False         True   

       is_senior  capital_net  capital_gain_log  
0              0     0.365112          0.138416  
1              0     0.070739         -0.157731  
2              0     0.070739         -0.157731  
3              0     0.070739         -0.157731  
4              0     0.070739         -0.157731  
...          ...          ...               ...  
32556          0     0.070739         -0.157731  
32557          0     0.070739         -0.157731  
32558          0     0.070739         -0.157731  
32559      

capital_gain has a mean of 1077 and median is 0 which is indicating that it is highly skewed and the max value is 99999 whcih emans that many people has low gain while few have very large.

So,applying log transformation brings the feature closer to normal distribution.    

4.Feature Selection:

Use the Isolation Forest algorithm to identify and remove outliers. Discuss how outliers can affect model performance.

In [60]:
from sklearn.ensemble import IsolationForest

# Use numeric data only for outlier detection
numeric_data = df_encoded.select_dtypes(include=['float64', 'int64'])

# Fit Isolation Forest
iso = IsolationForest(contamination=0.01, random_state=42)  # 1% contamination
outlier_pred = iso.fit_predict(numeric_data)

# Add outlier flag to the dataframe
df_encoded['outlier'] = outlier_pred

# Filter out the outliers
df_cleaned = df_encoded[df_encoded['outlier'] == 1].drop(columns=['outlier'])


How Outliers Affect Model Performance:

- Outliers can shift the decision boundary.
- Affects statistical assumptions in linear models.
- Leads to poor generalization and higher prediction error.
- Outliers distort cluster centers or class boundaries.

Apply the PPS (Predictive Power Score) to find and discuss the relationships between features. Compare its findings with the correlation matrix.

import ppscore as pps

pps_matrix = pps.matrix(df_cleaned)



In [104]:
df_cleaned.corr()

workclass    fnlwgt  education  marital_status  \
workclass                  1.000000 -0.016847   0.022679       -0.063747   
fnlwgt                    -0.016847  1.000000  -0.028835        0.028575   
education                  0.022679 -0.028835   1.000000       -0.037641   
marital_status            -0.063747  0.028575  -0.037641        1.000000   
occupation                 0.253445  0.000692  -0.022195       -0.008801   
relationship              -0.090589  0.008434  -0.009525        0.185289   
native_country            -0.008077 -0.053543   0.064910       -0.023639   
age_scaled                 0.001674 -0.077198  -0.011728       -0.267835   
fnlwgt_scaled             -0.016847  1.000000  -0.028835        0.028575   
education_num_scaled       0.050016 -0.044452   0.358853       -0.066828   
capital_gain_scaled        0.027532 -0.010085   0.025769       -0.065017   
capital_loss_scaled        0.011838 -0.016480   0.019036       -0.035395   
hours_per_week_scaled      0.137945 -0.020153   0.054426       -0.189870   
sex_ Male                  0.094904  0.026914  -0.028927       -0.128342   
income_ >50K               0.048071 -0.011874   0.077090       -0.198404   
race_ Asian-Pac-Islander   0.004756 -0.051956   0.010229        0.017985   
race_ Black               -0.059303  0.118001  -0.021357        0.081062   
race_ Other               -0.015202  0.007063  -0.029619        0.016854   
race_ White                0.056620 -0.056546   0.022020       -0.080063   
is_senior                       NaN       NaN        NaN             NaN   
capital_net               -0.003085  0.012565  -0.010379        0.014291   
capital_gain_log           0.024432 -0.010342   0.024936       -0.068487   

                          occupation  relationship  native_country  \
workclass                   0.253445     -0.090589       -0.008077   
fnlwgt                      0.000692      0.008434       -0.053543   
education                  -0.022195     -0.009525        0.064910   
marital_status             -0.008801      0.185289       -0.023639   
occupation                  1.000000     -0.074833       -0.012370   
relationship               -0.074833      1.000000       -0.004818   
native_country             -0.012370     -0.004818        1.000000   
age_scaled                 -0.021642     -0.263276       -0.001489   
fnlwgt_scaled               0.000692      0.008434       -0.053543   
education_num_scaled        0.107938     -0.090345        0.052014   
capital_gain_scaled         0.023506     -0.078412        0.010155   
capital_loss_scaled         0.017499     -0.059413       -0.001792   
hours_per_week_scaled       0.078212     -0.247018       -0.003373   
sex_ Male                   0.079173     -0.580919       -0.008463   
income_ >50K                0.074016     -0.248802        0.016616   
race_ Asian-Pac-Islander    0.003861      0.011556       -0.212053   
race_ Black                -0.011300      0.138168       -0.003896   
race_ Other                 0.001196      0.016420       -0.106797   
race_ White                 0.008156     -0.132223        0.130704   
is_senior                        NaN           NaN             NaN   
capital_net                -0.009595      0.032985        0.004661   
capital_gain_log            0.022209     -0.083739        0.010179   

                          age_scaled  fnlwgt_scaled  education_num_scaled  \
workclass                   0.001674      -0.016847              0.050016   
fnlwgt                     -0.077198       1.000000             -0.044452   
education                  -0.011728      -0.028835              0.358853   
marital_status             -0.267835       0.028575             -0.066828   
occupation                 -0.021642       0.000692              0.107938   
relationship               -0.263276       0.008434             -0.090345   
native_country             -0.001489      -0.053543              0.052014   
age_scaled                  1.000000      -0.07719

- capital_gain_log and capital_gain_scaled has the highest positive correlation.
- education_num_scaled and income_ >50K also has a positive correlation meaning the one with high education has income greater than 50K.
- capital_gain_log and income_ >50K also has a positive correlation meaning the one with high capital gain has income greater than 50K.


Due to package version conflict i was unable to install the ppscore package so i have just written the code for it which will generate result in compatible environment.